File: common/src/main/java/com/pluxity/facility/Facility.java
package com.pluxity.facility;

import com.pluxity.facility.category.FacilityCategory;
import com.pluxity.feature.entity.Feature;
import com.pluxity.file.entity.FileEntity;
import com.pluxity.global.entity.BaseEntity;
import jakarta.persistence.*;
import java.util.ArrayList;
import java.util.List;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.SoftDelete;
import org.hibernate.envers.Audited;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Entity
@Table(name = "facility")
@Inheritance(strategy = InheritanceType.JOINED)
@DiscriminatorColumn(name = "facility_type")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@EntityListeners(AuditingEntityListener.class)
@SoftDelete
public abstract class Facility extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    private FacilityCategory category;

    @Column(name = "code", unique = true, length = 10)
    private String code;

    @Audited
    @Column(name = "drawing_file_id")
    private Long drawingFileId;

    @Column(name = "thumbnail_file_id")
    private Long thumbnailFileId;

    @Column(name = "name", unique = true, nullable = false, length = 50)
    private String name;

    @Column(name = "description")
    private String description;

    @Column(name = "history_comment")
    private String historyComment;

    @OneToMany(mappedBy = "facility")
    private final List<Feature> features = new ArrayList<>();

    protected Facility(String name, String description) {
        this(name, null, description, null);
    }

    public Facility(String name, String description, String historyComment) {
        this(name, null, description, historyComment);
    }

    protected Facility(String name, String code, Long drawingFileId, Long thumbnailFileId) {
        this.code = code;
        this.name = name;
        this.drawingFileId = drawingFileId;
        this.thumbnailFileId = thumbnailFileId;
    }

    protected Facility(String name, String code, String description, String historyComment) {
        this.code = code;
        this.name = name;
        this.description = description;
        this.historyComment = historyComment;
    }

    public Facility(
            String name, String code, String description, Long drawingFileId, Long thumbnailFileId) {
        this.name = name;
        this.code = code;
        this.description = description;
        this.drawingFileId = drawingFileId;
        this.thumbnailFileId = thumbnailFileId;
    }

    public void updateDrawingFileId(FileEntity drawingFile) {
        this.drawingFileId = drawingFile.getId();
    }

    public void updateThumbnailFileId(FileEntity thumbnailFile) {
        this.thumbnailFileId = thumbnailFile.getId();
    }

    public void updateDrawingFileId(Long drawingFileId) {
        this.drawingFileId = drawingFileId;
    }

    public void updateThumbnailFileId(Long thumbnailFileId) {
        this.thumbnailFileId = thumbnailFileId;
    }

    public void updateName(String name) {
        this.name = name;
    }

    public void updateCode(String code) {
        this.code = code;
    }

    public void updateDescription(String description) {
        this.description = description;
    }

    public void updateHistoryComment(String historyComment) {
        this.historyComment = historyComment;
    }

    public void assignCategory(FacilityCategory category) {
        this.category = category;
    }

    public void addFeature(Feature feature) {
        if (!this.features.contains(feature)) {
            this.features.add(feature);
        }
    }

    public void removeFeature(Feature feature) {
        this.features.remove(feature);
    }

    public void update(Facility facility) {
        if (facility.name != null) {
            this.name = facility.name;
        }
        if (facility.code != null) {
            this.code = facility.code;
        }
        if (facility.description != null) {
            this.description = facility.description;
        }
        if (facility.historyComment != null) {
            this.historyComment = facility.historyComment;
        }
        if (facility.drawingFileId != null) {
            this.drawingFileId = facility.drawingFileId;
        }
        if (facility.thumbnailFileId != null) {
            this.thumbnailFileId = facility.thumbnailFileId;
        }
    }
}

File: common/src/main/kotlin/com/pluxity/facility/Facility.kt
package com.pluxity.facility

import com.pluxity.facility.category.FacilityCategory
import com.pluxity.facility.path.FacilityPath
import com.pluxity.feature.entity.Feature
import com.pluxity.file.entity.FileEntity
import com.pluxity.global.entity.BaseEntity
import com.pluxity.permission.ResourceType
import com.pluxity.user.entity.Permissible
import jakarta.persistence.Column
import jakarta.persistence.DiscriminatorColumn
import jakarta.persistence.Embedded
import jakarta.persistence.Entity
import jakarta.persistence.EntityListeners
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Inheritance
import jakarta.persistence.InheritanceType
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import org.hibernate.annotations.SoftDelete
import org.springframework.data.jpa.domain.support.AuditingEntityListener

@Entity
@Table(name = "facility")
@Inheritance(strategy = InheritanceType.JOINED)
@DiscriminatorColumn(name = "facility_type")
@EntityListeners(AuditingEntityListener::class)
@SoftDelete
abstract class Facility(
    @Column(name = "name", nullable = false, length = 50)
    open var name: String,
    @Column(name = "code", length = 50)
    open var code: String? = null,
    @Column(name = "description")
    open var description: String? = null,
    @Column(name = "history_comment")
    open var historyComment: String? = null,
    @Column(name = "drawing_file_id")
    open var drawingFileId: Long? = null,
    @Column(name = "thumbnail_file_id")
    open var thumbnailFileId: Long? = null,
    @Embedded
    open var position: FacilityPosition? = null,
) : BaseEntity(),
    Permissible {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open val id: Long? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "category_id")
    open var category: FacilityCategory? = null

    @Enumerated(EnumType.STRING)
    @Column(name = "facility_type", insertable = false, updatable = false)
    open var facilityType: FacilityType? = null

    @OneToMany(mappedBy = "facility")
    open val features: MutableList<Feature> = mutableListOf()

    @OneToMany(mappedBy = "facility")
    open val paths: MutableList<FacilityPath> = mutableListOf()

    fun requiredId(): Long = checkNotNull(this.id) { "Facility ID is required" }

    fun updateCode(code: String?) {
        this.code = code
    }

    fun updateName(name: String) {
        require(name.isNotBlank()) { "시설명은 빈 값일 수 없습니다" }
        this.name = name
    }

    fun updateDescription(description: String?) {
        this.description = description
    }

    fun updateDrawingFile(drawingFile: FileEntity) {
        this.drawingFileId = drawingFile.id
    }

    fun updateThumbnailFileId(thumbnailFileId: Long?) {
        this.thumbnailFileId = thumbnailFileId
    }

    fun updateThumbnailFile(thumbnailFile: FileEntity) {
        this.thumbnailFileId = thumbnailFile.id
    }

    fun updatePosition(
        lon: Double?,
        lat: Double?,
        locationMeta: String?,
    ) {
        this.position =
            if (lon != null || lat != null || locationMeta != null) {
                FacilityPosition(lon, lat, locationMeta)
            } else {
                null
            }
    }

    fun assignCategory(category: FacilityCategory) {
        this.category = category
    }

    fun update(other: Facility) {
        this.name = other.name
        this.code = other.code
        this.description = other.description
        this.historyComment = other.historyComment
        this.drawingFileId = other.drawingFileId
        this.thumbnailFileId = other.thumbnailFileId
        this.position = other.position
        this.category = other.category
    }

    override val resourceId: String
        get() = id.toString()

    override val resourceType: ResourceType
        get() = ResourceType.FACILITY
}
